File: apps/knowledge-service/src/services/integrity-checker.ts
import { readFile, readdir, stat, access } from "node:fs/promises";
import { join, dirname, relative, resolve, extname } from "node:path";

export interface DebtItem {
  type:
    | "missing_crosslink"
    | "missing_hub"
    | "orphan_doc"
    | "broken_link"
    | "stale_doc";
  path: string;
  description: string;
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

async function fileExists(p: string): Promise<boolean> {
  try {
    await access(p);
    return true;
  } catch {
    return false;
  }
}

/** Collect all .md files under a directory */
async function collectMd(dir: string): Promise<string[]> {
  const results: string[] = [];
  try {
    const entries = await readdir(dir, { withFileTypes: true });
    for (const entry of entries) {
      const full = join(dir, entry.name);
      if (entry.isDirectory()) {
        if (entry.name.startsWith(".") || entry.name === "node_modules")
          continue;
        results.push(...(await collectMd(full)));
      } else if (extname(entry.name) === ".md") {
        results.push(full);
      }
    }
  } catch {
    // skip
  }
  return results;
}

/** Extract markdown links [text](path) from content */
function extractLinks(content: string): string[] {
  const linkRe = /\[([^\]]*)\]\(([^)]+)\)/g;
  const links: string[] = [];
  let m: RegExpExecArray | null;
  while ((m = linkRe.exec(content)) !== null) {
    const href = m[2];
    // only relative links (no http, no anchors-only)
    if (!href.startsWith("http") && !href.startsWith("#")) {
      links.push(href.split("#")[0]); // strip anchor
    }
  }
  return links;
}

/** Check if content has a "## Related" section with at least one markdown link */
function hasCrosslinkSection(content: string): boolean {
  const relatedIdx = content.indexOf("## Related");
  if (relatedIdx === -1) return false;
  const after = content.slice(relatedIdx);
  return /\[([^\]]+)\]\(([^)]+)\)/.test(after);
}

/** Find hub/index docs in a folder */
async function findHubDoc(dir: string): Promise<string | null> {
  const candidates = ["index.md", "hub.md", "README.md", "_index.md"];
  for (const c of candidates) {
    const p = join(dir, c);
    if (await fileExists(p)) return p;
  }
  return null;
}

// ---------------------------------------------------------------------------
// Main integrity check
// ---------------------------------------------------------------------------
export async function checkIntegrity(
  modifiedDocs: string[],
  basePath: string,
): Promise<{ pass: boolean; debts: DebtItem[] }> {
  const debts: DebtItem[] = [];

  for (const doc of modifiedDocs) {
    const fullPath = resolve(basePath, doc);

    let content: string;
    try {
      content = await readFile(fullPath, "utf-8");
    } catch {
      // file doesn't exist or unreadable — skip
      continue;
    }

    // 1. Crosslink check
    if (!hasCrosslinkSection(content)) {
      debts.push({
        type: "missing_crosslink",
        path: doc,
        description: `Missing "## Related" section with crosslinks`,
      });
    }

    // 2. Hub registration check
    const folder = dirname(fullPath);
    const hubDoc = await findHubDoc(folder);
    if (hubDoc) {
      const hubContent = await readFile(hubDoc, "utf-8");
      const docBasename = fullPath.split("/").pop() ?? "";
      if (!hubContent.includes(docBasename)) {
        debts.push({
          type: "missing_hub",
          path: doc,
          description: `Not listed in hub doc ${relative(basePath, hubDoc)}`,
        });
      }
    }

    // 3. Link validation
    const links = extractLinks(content);
    for (const link of links) {
      const resolved = resolve(dirname(fullPath), link);
      if (!(await fileExists(resolved))) {
        debts.push({
          type: "broken_link",
          path: doc,
          description: `Broken link: ${link}`,
        });
      }
    }

    // 4. Staleness check (frontmatter)
    const frontmatterMatch = content.match(/^---\n([\s\S]*?)\n---/);
    if (frontmatterMatch) {
      const fm = frontmatterMatch[1];
      if (/status:\s*(deprecated|stale)/i.test(fm)) {
        debts.push({
          type: "stale_doc",
          path: doc,
          description: `Doc is marked as deprecated/stale in frontmatter`,
        });
      }
    }
  }

  // 5. Orphan detection — scan all docs in basePath, find ones not referenced by any hub
  try {
    const allDocs = await collectMd(basePath);
    for (const docPath of allDocs) {
      const folder = dirname(docPath);
      const hubDoc = await findHubDoc(folder);
      if (!hubDoc || hubDoc === docPath) continue;

      const hubContent = await readFile(hubDoc, "utf-8");
      const basename = docPath.split("/").pop() ?? "";
      // Skip hub docs themselves
      if (
        ["index.md", "hub.md", "README.md", "_index.md"].includes(basename)
      ) {
        continue;
      }

      if (!hubContent.includes(basename)) {
        debts.push({
          type: "orphan_doc",
          path: relative(basePath, docPath),
          description: `Orphan: not referenced by ${relative(basePath, hubDoc)}`,
        });
      }
    }
  } catch {
    // basePath may not be a valid directory
  }

  return { pass: debts.length === 0, debts };
}

File: knowledge/langgraph/knowledge/services/integrity_checker.py
"""Knowledge integrity checker -- ported from integrity-checker.ts."""

from __future__ import annotations

import os
import re
from pathlib import Path

from langgraph.knowledge.types import DebtItem

_HUB_CANDIDATES = ("index.md", "hub.md", "README.md", "_index.md")


def _collect_md(directory: str) -> list[str]:
    """Recursively collect ``.md`` files, skipping hidden dirs and node_modules."""

    results: list[str] = []
    try:
        for entry in os.scandir(directory):
            if entry.is_dir(follow_symlinks=False):
                if entry.name.startswith(".") or entry.name == "node_modules":
                    continue
                results.extend(_collect_md(entry.path))
            elif entry.is_file() and entry.name.endswith(".md"):
                results.append(entry.path)
    except OSError:
        pass
    return results


def _extract_links(content: str) -> list[str]:
    """Extract relative markdown links ``[text](path)``."""

    links: list[str] = []
    for m in re.finditer(r"\[([^\]]*)\]\(([^)]+)\)", content):
        href = m.group(2)
        if not href.startswith("http") and not href.startswith("#"):
            links.append(href.split("#")[0])
    return links


def _has_crosslink_section(content: str) -> bool:
    idx = content.find("## Related")
    if idx == -1:
        return False
    after = content[idx:]
    return bool(re.search(r"\[([^\]]+)\]\(([^)]+)\)", after))


def _find_hub_doc(directory: str) -> str | None:
    for name in _HUB_CANDIDATES:
        candidate = os.path.join(directory, name)
        if os.path.isfile(candidate):
            return candidate
    return None


async def check_integrity(
    modified_docs: list[str],
    base_path: str,
) -> tuple[bool, list[DebtItem]]:
    """Run integrity checks on *modified_docs* and return ``(pass, debts)``."""

    debts: list[DebtItem] = []

    for doc in modified_docs:
        full_path = os.path.normpath(os.path.join(base_path, doc))

        try:
            content = Path(full_path).read_text(encoding="utf-8")
        except OSError:
            continue

        # 1. Crosslink check
        if not _has_crosslink_section(content):
            debts.append(
                DebtItem(
                    type="missing_crosslink",
                    path=doc,
                    description='Missing "## Related" section with crosslinks',
                )
            )

        # 2. Hub registration check
        folder = os.path.dirname(full_path)
        hub_doc = _find_hub_doc(folder)
        if hub_doc is not None:
            hub_content = Path(hub_doc).read_text(encoding="utf-8")
            doc_basename = os.path.basename(full_path)
            if doc_basename not in hub_content:
                debts.append(
                    DebtItem(
                        type="missing_hub",
                        path=doc,
                        description=f"Not listed in hub doc {os.path.relpath(hub_doc, base_path)}",
                    )
                )

        # 3. Link validation
        for link in _extract_links(content):
            resolved = os.path.normpath(
                os.path.join(os.path.dirname(full_path), link)
            )
            if not os.path.exists(resolved):
                debts.append(
                    DebtItem(
                        type="broken_link",
                        path=doc,
                        description=f"Broken link: {link}",
                    )
                )

        # 4. Staleness check (frontmatter)
        fm_match = re.match(r"^---\n([\s\S]*?)\n---", content)
        if fm_match and re.search(r"status:\s*(deprecated|stale)", fm_match.group(1), re.IGNORECASE):
            debts.append(
                DebtItem(
                    type="stale_doc",
                    path=doc,
                    description="Doc is marked as deprecated/stale in frontmatter",
                )
            )

    # 5. Orphan detection
    try:
        all_docs = _collect_md(base_path)
        for doc_path in all_docs:
            folder = os.path.dirname(doc_path)
            hub_doc = _find_hub_doc(folder)
            if hub_doc is None or hub_doc == doc_path:
                continue

            basename = os.path.basename(doc_path)
            if basename in _HUB_CANDIDATES:
                continue

            hub_content = Path(hub_doc).read_text(encoding="utf-8")
            if basename not in hub_content:
                debts.append(
                    DebtItem(
                        type="orphan_doc",
                        path=os.path.relpath(doc_path, base_path),
                        description=f"Orphan: not referenced by {os.path.relpath(hub_doc, base_path)}",
                    )
                )
    except OSError:
        pass

    return (len(debts) == 0, debts)
